### ISSUE TITLE
Add drx_expand() function for VSIB AVX2/AVX-512 scatter/gather.

### ISSUE THREAD

**Issue by sdconsta** [DynamoRIO/dynamorio#2985]
When encountering the following instruction:
```
c4 e2 65 90 14 80       vpgatherdd %ymm3,(%rax,%ymm0,4),%ymm2
```
compiled from the intrinsic
```
__m256i _mm256_i32gather_epi32 (int const* base_addr, __m256i vindex, const int scale)
#include <immintrin.h>
Instruction: vpgatherdd ymm, vm32x, ymm
CPUID Flags: AVX2
```
with either `clang` or `gcc`. DynamoRIO complains that
```
ERROR: Could not find encoding for: lea    (%rax,%ymm0,4) -> %rbp
SYSLOG_ERROR: Application /home/scott/Documents/covert-cpp/build/bin/DynLoader (27816) DynamoRIO usage error : instr_encode error: no encoding found (see log)
SYSLOG_ERROR: Usage error: instr_encode error: no encoding found (see log) (/home/travis/build/DynamoRIO/dynamorio/core/arch/x86/encode.c, line 2423)
version 7.0.17598, custom build
```
It looks like DynamoRIO is expecting the base address to be 4 bytes, but the actual size is 8 bytes:
https://github.com/DynamoRIO/dynamorio/blob/01357f8ae547b4f4ea66d39a3214aed662428a0d/suite/tests/api/ir_x86_3args.h#L384.

**Comment by derekbruening** [comment]
I can't reproduce this.  Decoding:
```
    const byte bytes[] = { 0xc4, 0xe2, 0x65, 0x90, 0x14, 0x80 };
    disassemble_with_info(dcontext, (byte *)bytes, STDERR, false, true);
```
Results in:
```
 c4 e2 65 90 14 80    vpgatherdd (%rax,%ymm0,4)[4byte] %ymm3 -> %ymm2 %ymm3
```
With 7.0.17652 (HEAD at b9726d5fa76be2c3ffe34fcf48ad73cc1f92858c).

I also don't understand this comment:
> It looks like DynamoRIO is expecting the base address to be 4 bytes, but the actual size is 8 bytes:

The memory operand size is 4 bytes (for each vector fetch): why do you think it should be 8 bytes?

Looking at the code I do see some issues: the disasm names do not match the mnemonics for a couple of related instructions (vpgatherqd and vpgatherqq) which I will fix.  But I do not see how you are getting an incorrect decoding?

**Comment by fidget324** [comment]
> I also don't understand this comment

My mistake! I didn't understand VSIB addressing.

Ok, I pulled down the latest Cron build, and I'm still having the same issue. Here's the simplest example that can reproduce the crash:
```C
#include <immintrin.h>
__m256i res;
int main() {
  int arr[256] = {0};
  __m256i vindex = {0};
  res = _mm256_i32gather_epi32(arr, vindex, sizeof(int));
}
```
I then run:
```
$ gcc -mavx2 test.c -o gcc.out
$ drrun -debug -c libmemtrace_simple.so -- gcc.out
<Starting application gcc.out (9854)>
<Initial options = -no_dynamic_options -client_lib 'DynamoRIO-x86_64-Linux-7.0.17651-0/samples/bin64/libmemtrace_simple.so;0;' -code_api -stack_size 56K -signal_stack_size 32K -max_elide_jmp 0 -max_elide_call 0 -early_inject -emulate_brk -no_inline_ignored_syscalls -native_exec_default_list '' -no_native_exec_managed_code -no_indcall2direct >
<Paste into GDB to debug DynamoRIO clients:
set confirm off
add-symbol-file 'DynamoRIO-x86_64-Linux-7.0.17651-0/samples/bin64/libmemtrace_simple.so' 0x00007fdb59a2b310
add-symbol-file 'DynamoRIO-x86_64-Linux-7.0.17651-0/lib64/debug/libdynamorio.so' 0x00007fdbdda595f8
add-symbol-file 'DynamoRIO-x86_64-Linux-7.0.17651-0/ext/lib64/debug/libdrutil.so' 0x00007fdb59c306c0
add-symbol-file 'DynamoRIO-x86_64-Linux-7.0.17651-0/ext/lib64/debug/libdrmgr.so' 0x00007fdb59e38130
add-symbol-file 'DynamoRIO-x86_64-Linux-7.0.17651-0/ext/lib64/debug/libdrx.so' 0x00007fdb5a044200
add-symbol-file 'DynamoRIO-x86_64-Linux-7.0.17651-0/ext/lib64/debug/libdrreg.so' 0x00007fdb5a24db30
add-symbol-file '/lib/x86_64-linux-gnu/libc.so.6' 0x00007fdbdd42e2d0
add-symbol-file '/lib64/ld-linux-x86-64.so.2' 0x00007fdbdd1e2f10
>
Data file DynamoRIO-x86_64-Linux-7.0.17651-0/samples/bin64/memtrace.gcc.out.09854.0000.log created
<(1+x) Handling our fault in a TRY at 0x00007fdbddc9edc8>
<spurious rep/repne prefix @0x00007fdbdd816746 (f2 41 ff e3): >
<Application gcc.out (9854) DynamoRIO usage error : instr_encode error: no encoding found (see log)>
<Usage error: instr_encode error: no encoding found (see log) (/home/travis/build/DynamoRIO/dynamorio/core/arch/x86/encode.c, line 2423)
version 7.0.17651, custom build
-no_dynamic_options -client_lib 'DynamoRIO-x86_64-Linux-7.0.17651-0/samples/bin64/libmemtrace_simple.so;0;' -code_api -stack_size 56K -signal_stack_size 32K -max_elide_jmp 0 -max_elide_call 0 -early_inject -emulate_brk -no_inline_ignored_syscalls -native_exec_default_list '' -no_native_exec_managed_
0x00007fdbc98da530 0x00007fdbddaf4b51
0x00007fdbc98da590 0x00007fdbddc5598a
0x00007fdbc98da6b0 0x00007fdbddc4b14f
0x00007fdbc98da6f0 0x00007fdbddc36082
0x00007fdbc98da750 0x00007fdbddc37bb2
0x00007fdbc98da780 0x00007fdbddaedb99
0x00007fdbc98dacd0 0x00007fdbddaf4561
0x00007fdbc98dad20 0x00007fdbddc7228f
0x00007fdbc98daf10 0x00007fdbddadd410
0x00007fdbc98daff0 0x00007fdbc9860ec5
0x00007ffe35baf930 0x00007fdbdcbdcb97
0x00007fdbd9827730 0x0676258d4c544155
DynamoRIO-x86_64-Linux-7.0.17651-0/samples/bin64/libmemtrace_simple.so=0x00007fdb59a29000
/lib/x86_64-linux-gnu/libc.so.6=0x00007fdbdd40d000
/lib64/ld-linux-x86-64.so.2=0x00007fdbdd1e2000
DynamoRIO-x86_64-Linux-7.0.17651-0/ext/lib64/debug/libdrx.so=0x00007fdb5a041000
DynamoRIO-x86_64-Linux-7.0.17651-0/ext/lib64/debug/libdrreg.so=0x00007fdb5a24b000
DynamoRIO-x>
```
I observe the exact same result when I compile with `clang`.

**Comment by derekbruening** [comment]
Well, that is a pretty important detail not in the initial filing (!!): you're running under memtrace_simple.  I assume that plain DR (i.e., `drrun -debug -- gcc.out`) works fine, right?

So this sounds like an instrumentation issue with acquiring the address of a VSIB instruction.

**Comment by fidget324** [comment]
Actually I was not testing with memtrace_simple when I made the initial filing. I was working with an internal tool which uses the DynamoRIO APIs. I just used memtrace_simple in my second comment to demonstrate the crash.

**Comment by fidget324** [comment]
And BTW I would appreciate any advice you might have on how to instrument an instruction that uses VSIB addressing. Is `instr_compute_address()` the right API? It looks like I'll need to do this for my tool!

**Comment by derekbruening** [comment]
To support the `drutil_insert_get_mem_addr()` usage model, one solution is to follow the approach we use for rep string instructions (via `drutil_expand_rep_string()`): rewrite the VSIB instruction into a series of instructions, facilitating simpler instrumentation where each memory reference is in its own separate instruction.

To keep it a single VSIB instruction, instrumentation would most easily call out to a slowpath in a clean call that used `instr_compute_address_ex()` to compute each dynamic address.  `instr_compute_address_ex()` already has support for VSIB.  Handling VSIB inline could be done if the slowpath becomes a bottleneck but sounds like a pain.

**Comment by fidget324** [comment]
> To support the drutil_insert_get_mem_addr() usage model, one solution is to follow the approach we use for rep string instructions (via drutil_expand_rep_string()): rewrite the VSIB instruction into a series of instructions, facilitating simpler instrumentation where each memory reference is in its own separate instruction.

This would be awesome.

> To keep it a single VSIB instruction, instrumentation would most easily call out to a slowpath in a clean call that used instr_compute_address_ex() to compute each dynamic address. instr_compute_address_ex() already has support for VSIB.

I've implemented a hack using a clean call, as you suggested, and it works! (but is quite slow):
```C
void vsib_mem_callback(instr_t *instr, int size) {
  dr_mcontext_t mc = {sizeof(mc),
                      DR_MC_CONTROL | DR_MC_INTEGER | DR_MC_MULTIMEDIA};
  dr_get_mcontext(dr_get_current_drcontext(), &mc);
  int i = 0;
  app_pc addr;
  bool is_write;
  while (instr_compute_address_ex(instr, &mc, i++, &addr, &is_write)) {
    dr_fprintf(STDERR, "Address: %p, Size: %d, Type: %c\n", addr, size, is_write ? 'w' : 'r');
  }
}

static void instrument_vsib_mem(void *drcontext, instrlist_t *bb,
                                instr_t *where, opnd_t ref) {
  instr_t *instr = instr_clone(drcontext, where);
  int opnd_sz = drutil_opnd_mem_size_in_bytes(ref, instr);
  dr_insert_clean_call(drcontext, bb, where, vsib_mem_callback, false, 2,
                       OPND_CREATE_INTPTR(instr), OPND_CREATE_INT32(opnd_sz));
}

static void instrument_instr(void *drcontext, instrlist_t *bb, instr_t *where) {
  if (instr_reads_memory(where)) {
    for (int i = 0; i < instr_num_srcs(where); i++) {
      opnd_t opnd = instr_get_src(where, i);
      if (opnd_is_vsib(opnd)) {
          instrument_vsib_mem(drcontext, bb, where, opnd);
      }
    }
  }
}
```
Hopefully this will be of use to anyone else looking for an easy solution...

> Handling VSIB inline could be done if the slowpath becomes a bottleneck but sounds like a pain.

I agree.

**Comment by toshipiazza** [comment]
> To support the drutil_insert_get_mem_addr() usage model, one solution is to follow the approach we use for rep string instructions (via drutil_expand_rep_string()): rewrite the VSIB instruction into a series of instructions, facilitating simpler instrumentation where each memory reference is in its own separate instruction.

We make a similar assumption in https://github.com/DynamoRIO/dynamorio/blob/master/api/samples/memval_simple.c#L301, looks like something like this where we expand the VSIB instruction would work for memval_simple as well.

**Comment by hgreving2304** [comment]
Do I understand this right that the issue here now is to support scatter/gather expansion drutil_expand_scatter_gather() (AVX2 only gather, but AVX-512 will have a scatter instruction) to be used in  drutil_insert_get_mem_addr()?

**Comment by derekbruening** [comment]
> to be used in drutil_insert_get_mem_addr()? 

I wouldn't say it would be used "in" drutil_insert_get_mem_addr(): like drutil_expand_rep_string() it would be a separate function called during the app2app stage.

**Comment by hgreving2304** [comment]
That's what I meant. Ok.

**Comment by hgreving2304** [comment]
It looks like that the sequence we can expand gather/scatter to is more simple than previously assumed. I first thought we need to add tests of the EVEX mask bit with _kmov_, _and_, followed by a _jne_ to conditionally skip the scalar load/store. It looks like that AVX-512 does support fault suppression both for loads and stores. This is not very well described in the Software manuals. The unaligned vector load/store support the exception class "E4", described in "Table 2-43. EVEX Instructions in each Exception Class".

For AVX-512 gather with write mask kN we can then use a linear sequence of vmovups/vmovupd/vmovdqu16/32/64 with the already present merge mask, simply accumulating the individual elements without introducing a transparency violating fault.

For AVX-512 scatter, the same applies.

For AVX2 gather, we can use vmaskmovps/pd the same way. AVX2 also supports fault suppression.

**Comment by derekbruening** [comment]
Don't forget to add the emulation markers from #3112: I don't see them there in the finished-so-far code.

**Comment by hgreving2304** [comment]
Yes, I saw Assad's comment and noted that. Need to sort this out (need to understand it better, talking to myself). We had talked offline mentioning restore needs to happen in drreg, but then I think at least a part of that needs to be in drx using the markers. WIP.

**Comment by petervbraun** [comment]
I believe I am having the same problem. Can I get some more information on where to implement your hack @sdconsta ? The signature for instrument_instr differs from what you have in the snippet. I'm using the latest source. Thanks!

**Comment by sdconsta** [comment]
All of the code that I wrote that handles VSIB operands is in this source file:

https://github.com/sdconsta/covert-cpp/blob/master/tools/nvt/NVT/NVTMain.c

Just search for "VSIB" and you'll find everything. I use instr_compute_address_ex() to iterate over the addresses encoded in the operand. Does this help?

On Tue, Jan 21, 2020 at 12:53 PM petervbraun <notifications@github.com<mailto:notifications@github.com>> wrote:

I believe I am having the same problem. Can I get some more information on where to implement your hack @sdconsta<https://github.com/sdconsta> ? The signature for instrument_instr differs from what you have in the snippet. I'm using the latest source. Thanks!

**Comment by prasun3** [comment]
@hgreving2304 -- I see a number of commits are merged already. What changes are still pending to complete this?

**Comment by hgreving2304** [comment]
@prasun3 if it's related to the scatter/gather expansion sequences (emulation to be able to instrument addresses), then I am not sure, these sequences don't explicitly generate lea. Please check with the active DR community if there is a fix in pending (e.g. mailing list dynamorio-devs).

**Comment by derekbruening** [comment]
@hgreving2304 if you could help us figure out whether the scatter-gather expansion support for AVX2/AVX-512 is fully complete or not?  Are there still any pieces missing or did your last commit in November finish it off?  Definitely the various tracing tools like drcachesim's drmemtrace tracer are not yet using this expansion code and so end up failing on scatter-gather instructions.  Is drx_expand_scatter_gather() ready to be used by drmemtrace?

**Comment by hgreving2304** [comment]
@derekbruening should have been functional and complete for 64-bit. The missing parts are pointed out in FIXMEs in drx: some 32-bit support, and automatic xmm spilling (I think @johnfxgalea was working on that last year). I would not expect the lea error above, this sounds like a bug, if it is related.

**Comment by derekbruening** [comment]
> I would not expect the lea error above, this sounds like a bug, if it is related.

The error is not coming from the scatter-gather expansion but rather a lack of it.  The error `Could not find encoding for: lea     (%rax,%ymm1,8) -> %rdx` comes from drmemtrace asking drutil_insert_get_mem_addr_ex() to obtain the address.  drmemtrace needs to instead invoke drx_expand_scatter_gather() during the app2app phase, just like it does with drutil_expand_rep_string_ex().  Plus, that drutil routine should have a check for scatter-gather and gracefully fail, rather than trying to insert un-encodable instructions.

**Comment by hgreving2304** [comment]
> The error is not coming from the scatter-gather expansion but rather a lack of it
Ok.

**Comment by derekbruening** [comment]
> The missing parts are pointed out in FIXMEs in drx: some 32-bit support, and automatic xmm spilling (I think @johnfxgalea was working on that last year).

The xmm spilling: looks like #3844 and there is a PR #3893 but it is not fully finished.

**Comment by johnfxgalea** [comment]
Yep, that PR is on my radar. Functionality-wise, all is implemented. However, to get it merged, I need to (perhaps heavily) refactor drreg and separate code for gpr and simd spill handling into two source files (I still need to see if flags also merit a separate file). Tieing up loose ends atm, but hopefully will have some time soon.

**Comment by abhinav92003** [comment]
We hit the `lea` encoding issue on another application. Based on [Derek's comment above](https://github.com/DynamoRIO/dynamorio/issues/2985#issuecomment-703722537), we need some fix in drmemtrace. If needed, I can help with that. But I'm not sure if the open PR #3893 needs to merge first. Just wanted to stir this conversation to see if it can be submitted soon, and if there are any existing plans for the `lea` encoding fix.

**Comment by johnfxgalea** [comment]
@abhinav92003 Just to let you know that progress is being made on this issue. I'll try to make a PR sometime this week.

**Comment by abhinav92003** [comment]
Trying to summarise the pending items for completing scatter-gather support, and their status:

- [ ] Invoking `drx_expand_scatter_gather` in drcachesim - WIP PR #4863 @abhinav92003 

- [ ] Issue #3823: Handle `drreg_reserve_register` usages in multiple instrumentation passes, and avoid spill slot conflict. WIP PR #4863 @abhinav92003 

- [ ] Issue #3844: SIMD spill support in drreg: WIP PR #3893 @johnfxgalea 

- [ ] Invoking `drx_expand_scatter_gather` in sample clients (`memval_simple`, `memtrace_simple`, `memtrace_x86`) -- I tried this but seeing some issues. WIP, probably will be part of PR #4863. @abhinav92003 

- [ ] drcachesim instrumented fragment of expanded gather instr doesn’t look right. Only the first read in the gather sequence has drcachesim instrumentation. Needs more investigation https://github.com/DynamoRIO/dynamorio/pull/4863#issuecomment-824560991

- [ ] X86 32-bit support for scatter-gather variations that have qword index or value - Not started yet

- [ ] Issue #3801: probably not a blocker as it's listed as low priority. Haven't look in detail yet.

**Comment by abhinav92003** [comment]
Another one split from one of the above:
- [ ] Issue #3823: Handle multi-phase reservation for aflags in drreg @abhinav92003

**Comment by abhinav92003** [comment]
On X86-64, 84bf9288a12e8e38abab04d4d8273cc3226fa13c was used to successfully gather drcachesim traces of an application with millions of scatter/gather instructions. So I believe that we have added sufficient support and solved all blocking stability problems to close this issue.

The few non-blocking items that are left from my above comment (https://github.com/DynamoRIO/dynamorio/issues/2985#issuecomment-826247798) are:


Features:
#5259 expand scatter/gather instrs with qword index or value on X86-32
#3844 drreg support for SIMD spills

Stability:
#4629  Indeterministic unsupported EVEX encoding failures in some tests.
#5005  Audit state restore for scatter/gather

Further progress can be tracked on the above issues.

I'm happy to finally close this 3.5 year old feature request :)! Huge thanks to everyone who worked on adding the required support. @hgreving2304 @johnfxgalea @derekbruening  (I hope I didn't miss anyone!)

**Comment by derekbruening** [comment]
There is also arm #3837 and aarch64 #5036
